File: src/harness_engineering/github_tracker.py
from __future__ import annotations

import json
import re
import urllib.error
import urllib.request
from dataclasses import dataclass
from datetime import datetime
from typing import Any, Protocol

from harness_engineering.config import TrackerConfig
from harness_engineering.models import Issue


class TrackerError(RuntimeError):
    def __init__(self, code: str, message: str, *, payload: Any | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.payload = payload


class GraphQLTransport(Protocol):
    def execute(self, query: str, variables: dict[str, Any], *, endpoint: str, api_key: str) -> dict[str, Any]: ...


@dataclass(slots=True)
class UrllibGraphQLTransport:
    timeout_seconds: float = 30.0

    def execute(self, query: str, variables: dict[str, Any], *, endpoint: str, api_key: str) -> dict[str, Any]:
        body = json.dumps({"query": query, "variables": variables}).encode("utf-8")
        request = urllib.request.Request(
            endpoint,
            data=body,
            method="POST",
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
                "Accept": "application/json",
                "User-Agent": "harness-engineering/0.1",
            },
        )
        try:
            with urllib.request.urlopen(request, timeout=self.timeout_seconds) as response:
                response_body = response.read()
        except urllib.error.HTTPError as exc:
            raise TrackerError("github_api_status", f"GitHub API returned HTTP {exc.code}") from exc
        except urllib.error.URLError as exc:
            raise TrackerError("github_api_request", f"GitHub API request failed: {exc.reason}") from exc
        try:
            decoded = json.loads(response_body.decode("utf-8"))
        except json.JSONDecodeError as exc:
            raise TrackerError("github_unknown_payload", "GitHub API returned invalid JSON") from exc
        if decoded.get("errors"):
            raise TrackerError("github_graphql_errors", "GitHub GraphQL response contained errors", payload=decoded)
        return decoded


class GitHubTracker:
    def __init__(self, config: TrackerConfig, *, transport: GraphQLTransport | None = None, page_size: int = 50) -> None:
        self.config = config
        self.transport = transport or UrllibGraphQLTransport()
        self.page_size = page_size

    def fetch_candidate_issues(self) -> list[Issue]:
        return self._fetch_by_states(self.config.active_states)

    def fetch_issues_by_states(self, state_names: list[str]) -> list[Issue]:
        if not state_names:
            return []
        return self._fetch_by_states([state.lower() for state in state_names])

    def fetch_issue_states_by_ids(self, issue_ids: list[str]) -> list[Issue]:
        if not issue_ids:
            return []
        payload = self._execute(
            _ISSUES_BY_ID_QUERY,
            {"ids": issue_ids},
        )
        nodes = payload.get("data", {}).get("nodes")
        if not isinstance(nodes, list):
            raise TrackerError("github_unknown_payload", "missing nodes in issue state response", payload=payload)
        return [self._normalize_issue_node(node) for node in nodes if isinstance(node, dict) and node.get("__typename") in {None, "Issue"}]

    def _fetch_by_states(self, states: list[str]) -> list[Issue]:
        issues: list[Issue] = []
        after: str | None = None
        graphql_states = [_github_state(state) for state in states]
        while True:
            payload = self._execute(
                _CANDIDATE_QUERY,
                {
                    "owner": self.config.owner,
                    "repo": self.config.repo,
                    "states": graphql_states,
                    "first": self.page_size,
                    "after": after,
                },
            )
            issue_connection = payload.get("data", {}).get("repository", {}).get("issues")
            if not isinstance(issue_connection, dict):
                raise TrackerError("github_unknown_payload", "missing repository.issues in GitHub response", payload=payload)
            nodes = issue_connection.get("nodes")
            if not isinstance(nodes, list):
                raise TrackerError("github_unknown_payload", "missing issue nodes in GitHub response", payload=payload)
            issues.extend(self._normalize_issue_node(node) for node in nodes if isinstance(node, dict))
            page_info = issue_connection.get("pageInfo")
            if not isinstance(page_info, dict):
                raise TrackerError("github_unknown_payload", "missing pageInfo in GitHub response", payload=payload)
            if not page_info.get("hasNextPage"):
                return issues
            after = page_info.get("endCursor")
            if not after:
                raise TrackerError("github_missing_end_cursor", "GitHub pagination reported next page without endCursor")

    def _execute(self, query: str, variables: dict[str, Any]) -> dict[str, Any]:
        if not self.config.api_key:
            raise TrackerError("missing_tracker_api_key", "GitHub API key is required")
        payload = self.transport.execute(query, variables, endpoint=self.config.endpoint, api_key=self.config.api_key)
        if payload.get("errors"):
            raise TrackerError("github_graphql_errors", "GitHub GraphQL response contained errors", payload=payload)
        return payload

    def _normalize_issue_node(self, node: dict[str, Any]) -> Issue:
        number = node.get("number")
        labels = [
            str(label.get("name")).lower()
            for label in node.get("labels", {}).get("nodes", [])
            if isinstance(label, dict) and label.get("name")
        ]
        return Issue(
            id=str(node.get("id") or ""),
            identifier=f"{self.config.repo}#{number}",
            title=str(node.get("title") or ""),
            description=node.get("body") or None,
            priority=_priority_from_labels(labels),
            state=str(node.get("state") or "").lower(),
            branch_name=None,
            url=node.get("url") or None,
            labels=labels,
            blocked_by=[],
            created_at=_parse_datetime(node.get("createdAt")),
            updated_at=_parse_datetime(node.get("updatedAt")),
        )


def _github_state(state: str) -> str:
    normalized = state.lower()
    if normalized in {"open", "todo", "in progress", "in_progress"}:
        return "OPEN"
    if normalized in {"closed", "done", "cancelled", "canceled", "duplicate"}:
        return "CLOSED"
    return state.upper()


def _priority_from_labels(labels: list[str]) -> int | None:
    for label in labels:
        match = re.search(r"(?:^priority[:\s-]*|^p)(\d+)$", label)
        if match:
            return int(match.group(1))
    return None


def _parse_datetime(value: Any) -> datetime | None:
    if not value:
        return None
    try:
        return datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    except ValueError:
        return None


_CANDIDATE_QUERY = """
query HarnessCandidateIssues($owner: String!, $repo: String!, $states: [IssueState!], $first: Int!, $after: String) {
  repository(owner: $owner, name: $repo) {
    issues(first: $first, after: $after, states: $states, orderBy: {field: CREATED_AT, direction: ASC}) {
      nodes {
        id
        number
        title
        body
        state
        url
        createdAt
        updatedAt
        labels(first: 50) {
          nodes { name }
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
""".strip()


_ISSUES_BY_ID_QUERY = """
query HarnessIssuesById($ids: [ID!]!) {
  nodes(ids: $ids) {
    __typename
    ... on Issue {
      id
      number
      title
      body
      state
      url
      createdAt
      updatedAt
      labels(first: 50) {
        nodes { name }
      }
    }
  }
}
""".strip()

File: lib/harness_engineering/github_tracker.ex
defmodule HarnessEngineering.GitHubTracker.Error do
  @moduledoc false

  defexception [:code, :message, :payload]

  @impl true
  def exception(opts) do
    %__MODULE__{
      code: Keyword.fetch!(opts, :code),
      message: Keyword.fetch!(opts, :message),
      payload: Keyword.get(opts, :payload)
    }
  end
end

defmodule HarnessEngineering.GitHubTracker.FixtureTransport do
  @moduledoc false

  use Agent

  alias HarnessEngineering.GitHubTracker.Error

  def start_link(responses) when is_list(responses) do
    Agent.start_link(fn -> %{responses: responses, calls: []} end)
  end

  def from_file(path) do
    with {:ok, body} <- File.read(path),
         {:ok, fixture} <- decode_json(body),
         {:ok, responses} <- fixture_responses(fixture),
         {:ok, pid} <- start_link(responses) do
      {:ok, %{transport: {__MODULE__, pid}, fixture: fixture, pid: pid}}
    else
      {:error, %Error{} = error} ->
        {:error, error}

      {:error, reason} ->
        {:error,
         %Error{code: "fixture_load_failed", message: "fixture load failed: #{inspect(reason)}"}}
    end
  end

  def execute(pid, query, variables, _opts) do
    Agent.get_and_update(pid, fn state ->
      case state.responses do
        [response | rest] ->
          call = %{"query" => query, "variables" => variables}
          {{:ok, response}, %{state | responses: rest, calls: state.calls ++ [call]}}

        [] ->
          error = %Error{
            code: "fixture_exhausted",
            message: "GitHub fixture transport has no responses left"
          }

          {{:error, error}, state}
      end
    end)
  end

  def calls(pid) do
    Agent.get(pid, & &1.calls)
  end

  def stop(pid) do
    Agent.stop(pid)
  end

  def decode_json(body) do
    {:ok, body |> :json.decode() |> normalize_json()}
  rescue
    _error ->
      {:error,
       %Error{code: "fixture_parse_error", message: "fixture file must contain valid JSON"}}
  end

  defp normalize_json(:null), do: nil

  defp normalize_json(value) when is_list(value), do: Enum.map(value, &normalize_json/1)

  defp normalize_json(value) when is_map(value) do
    value
    |> Enum.map(fn {key, item} -> {to_string(key), normalize_json(item)} end)
    |> Map.new()
  end

  defp normalize_json(value), do: value

  defp fixture_responses(value) when is_list(value), do: {:ok, value}

  defp fixture_responses(%{"responses" => responses}) when is_list(responses),
    do: {:ok, responses}

  defp fixture_responses(_value) do
    {:error,
     %Error{
       code: "fixture_missing_responses",
       message: "fixture must be a response array or contain responses[]"
     }}
  end
end

defmodule HarnessEngineering.GitHubTracker do
  @moduledoc """
  GitHub issue tracker adapter for the Elixir port.

  The transport is injected so tests and one-shot dry runs can use captured
  GraphQL responses without network access.
  """

  alias HarnessEngineering.Config.Tracker
  alias HarnessEngineering.GitHubTracker.Error
  alias HarnessEngineering.Models.BlockerRef
  alias HarnessEngineering.Models.Issue

  @candidate_query """
  query HarnessCandidateIssues($owner: String!, $repo: String!, $states: [IssueState!], $first: Int!, $after: String) {
    repository(owner: $owner, name: $repo) {
      issues(first: $first, after: $after, states: $states, orderBy: {field: CREATED_AT, direction: ASC}) {
        nodes {
          id
          number
          title
          body
          state
          url
          createdAt
          updatedAt
          labels(first: 50) {
            nodes { name }
          }
        }
        pageInfo {
          hasNextPage
          endCursor
        }
      }
    }
  }
  """

  def fetch_candidate_issues(%Tracker{} = config, transport, page_size \\ 50) do
    fetch_by_states(config, transport, config.active_states, page_size)
  end

  def fetch_by_states(_config, _transport, [], _page_size), do: {:ok, []}

  def fetch_by_states(%Tracker{} = config, transport, states, page_size) do
    graphql_states = Enum.map(states, &github_state/1)
    do_fetch_by_states(config, transport, graphql_states, page_size, nil, [])
  end

  def normalize_issue_node(node, %Tracker{} = config) when is_map(node) do
    labels =
      node
      |> get_in(["labels", "nodes"])
      |> case do
        values when is_list(values) -> values
        _ -> []
      end
      |> Enum.flat_map(fn
        %{"name" => name} when not is_nil(name) -> [name |> to_string() |> String.downcase()]
        _other -> []
      end)

    number = Map.get(node, "number")
    repo = config.repo || ""

    %Issue{
      id: string_value(Map.get(node, "id")),
      identifier: "#{repo}##{number}",
      title: string_value(Map.get(node, "title")),
      description: optional_string(Map.get(node, "body")),
      priority: priority_from_labels(labels),
      state: node |> Map.get("state") |> string_value() |> String.downcase(),
      branch_name: nil,
      url: optional_string(Map.get(node, "url")),
      labels: labels,
      blocked_by: blocked_refs(node, repo),
      created_at: parse_datetime(Map.get(node, "createdAt")),
      updated_at: parse_datetime(Map.get(node, "updatedAt"))
    }
  end

  defp do_fetch_by_states(config, transport, graphql_states, page_size, after_cursor, acc) do
    variables = %{
      "owner" => config.owner,
      "repo" => config.repo,
      "states" => graphql_states,
      "first" => page_size,
      "after" => after_cursor
    }

    with {:ok, payload} <-
           execute(transport, @candidate_query, variables,
             endpoint: config.endpoint,
             api_key: config.api_key
           ),
         {:ok, connection} <- issue_connection(payload),
         {:ok, nodes} <- issue_nodes(connection),
         {:ok, page_info} <- page_info(connection) do
      issues =
        nodes
        |> Enum.filter(&is_map/1)
        |> Enum.map(&normalize_issue_node(&1, config))

      if Map.get(page_info, "hasNextPage") do
        case Map.get(page_info, "endCursor") do
          nil ->
            {:error,
             %Error{
               code: "github_missing_end_cursor",
               message: "GitHub pagination reported next page without endCursor"
             }}

          cursor ->
            do_fetch_by_states(
              config,
              transport,
              graphql_states,
              page_size,
              cursor,
              acc ++ issues
            )
        end
      else
        {:ok, acc ++ issues}
      end
    end
  end

  defp execute({module, ref}, query, variables, opts),
    do: module.execute(ref, query, variables, opts)

  defp execute(module, query, variables, opts) when is_atom(module),
    do: module.execute(query, variables, opts)

  defp issue_connection(payload) do
    case get_in(payload, ["data", "repository", "issues"]) do
      value when is_map(value) ->
        {:ok, value}

      _ ->
        {:error,
         %Error{
           code: "github_unknown_payload",
           message: "missing repository.issues in GitHub response",
           payload: payload
         }}
    end
  end

  defp issue_nodes(connection) do
    case Map.get(connection, "nodes") do
      value when is_list(value) ->
        {:ok, value}

      _ ->
        {:error,
         %Error{
           code: "github_unknown_payload",
           message: "missing issue nodes in GitHub response",
           payload: connection
         }}
    end
  end

  defp page_info(connection) do
    case Map.get(connection, "pageInfo") do
      value when is_map(value) ->
        {:ok, value}

      _ ->
        {:error,
         %Error{
           code: "github_unknown_payload",
           message: "missing pageInfo in GitHub response",
           payload: connection
         }}
    end
  end

  defp github_state(state) do
    case state |> to_string() |> String.downcase() do
      value when value in ["open", "todo", "in progress", "in_progress"] -> "OPEN"
      value when value in ["closed", "done", "cancelled", "canceled", "duplicate"] -> "CLOSED"
      value -> String.upcase(value)
    end
  end

  defp priority_from_labels(labels) do
    Enum.find_value(labels, fn label ->
      case Regex.run(~r/(?:^priority[:\s-]*|^p)(\d+)$/, label) do
        [_match, value] -> String.to_integer(value)
        _ -> nil
      end
    end)
  end

  defp blocked_refs(node, repo) do
    node_refs =
      node
      |> Map.get("blockedBy")
      |> blocked_nodes()
      |> Enum.map(&blocked_ref_from_node(&1, repo))

    body_refs =
      node
      |> Map.get("body")
      |> blocked_refs_from_body(repo)

    (node_refs ++ body_refs)
    |> Enum.uniq_by(fn blocker -> {blocker.id, blocker.identifier} end)
  end

  defp blocked_nodes(%{"nodes" => nodes}) when is_list(nodes), do: nodes
  defp blocked_nodes(nodes) when is_list(nodes), do: nodes
  defp blocked_nodes(_value), do: []

  defp blocked_ref_from_node(node, repo) do
    number = Map.get(node, "number")
    identifier = Map.get(node, "identifier") || if number, do: "#{repo}##{number}", else: nil

    %BlockerRef{
      id: optional_string(Map.get(node, "id")),
      identifier: optional_string(identifier),
      state: node |> Map.get("state") |> optional_downcase()
    }
  end

  defp blocked_refs_from_body(nil, _repo), do: []

  defp blocked_refs_from_body(body, repo) do
    case Regex.run(~r/(?is)^##\s*Blocked by\s*(.*?)(?:^##\s+|\z)/m, to_string(body),
           capture: :all_but_first
         ) do
      [section] ->
        Regex.scan(~r/#(\d+)/, section)
        |> Enum.map(fn [_match, number] -> %BlockerRef{identifier: "#{repo}##{number}"} end)

      _ ->
        []
    end
  end

  defp parse_datetime(nil), do: nil

  defp parse_datetime(value) do
    value
    |> to_string()
    |> DateTime.from_iso8601()
    |> case do
      {:ok, datetime, _offset} -> datetime
      _ -> nil
    end
  end

  defp string_value(nil), do: ""
  defp string_value(value), do: to_string(value)

  defp optional_string(nil), do: nil
  defp optional_string(""), do: nil
  defp optional_string(value), do: to_string(value)

  defp optional_downcase(nil), do: nil
  defp optional_downcase(value), do: value |> to_string() |> String.downcase()
end
